src/ucp/amo/nb_amo.c

```diff
@@ -13,24 +13,191 @@
 #include <ucs/debug/log.h>
 #include <inttypes.h>
 
+#define UCP_UCT_AMO_OPCODE_POST(_opcode, _op)                 \
+    if (ucs_unlikely((_opcode) >= UCP_ATOMIC_POST_OP_LAST)) { \
+        return UCS_ERR_INVALID_PARAM;                         \
+    } else {                                                  \
+        _op = ucp_uct_op_table[_opcode];                      \
+    }
+
+#define UCP_UCT_AMO_OPCODE_FETCH(_opcode, _op)                 \
+    if (ucs_unlikely((_opcode) >= UCP_ATOMIC_FETCH_OP_LAST)) { \
+        return UCS_ERR_INVALID_PARAM;                          \
+    } else {                                                   \
+        _op = ucp_uct_fop_table[_opcode];                      \
+    }
+
+static uct_atomic_op_t ucp_uct_op_table[] = {
+    [UCP_ATOMIC_POST_OP_ADD]    = UCT_ATOMIC_OP_ADD,
+    [UCP_ATOMIC_POST_OP_AND]    = UCT_ATOMIC_OP_AND,
+    [UCP_ATOMIC_POST_OP_OR]     = UCT_ATOMIC_OP_OR,
+    [UCP_ATOMIC_POST_OP_XOR]    = UCT_ATOMIC_OP_XOR
+};
+
+static uct_atomic_op_t ucp_uct_fop_table[] = {
+    [UCP_ATOMIC_FETCH_OP_FADD]  = UCT_ATOMIC_OP_ADD,
+    [UCP_ATOMIC_FETCH_OP_FAND]  = UCT_ATOMIC_OP_AND,
+    [UCP_ATOMIC_FETCH_OP_FOR]   = UCT_ATOMIC_OP_OR,
+    [UCP_ATOMIC_FETCH_OP_FXOR]  = UCT_ATOMIC_OP_XOR,
+    [UCP_ATOMIC_FETCH_OP_SWAP]  = UCT_ATOMIC_OP_SWAP,
+    [UCP_ATOMIC_FETCH_OP_CSWAP] = UCT_ATOMIC_OP_CSWAP,
+};
+
+
+static void ucp_amo_completed_single(uct_completion_t *self,
+                                     ucs_status_t status)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
+    ucs_trace("Invoking completion on AMO request %p", req);
```

Comment:
pls start message with smallcase letter

Reply:
hmmm, I didn't change this code (just shifted it from amo.inl file)